### ISSUE TITLE
Relationship to SRI

### ISSUE THREAD

**Issue by martinthomson** [httpwg/http-extensions#868]
It would be awesome if this could relate directly to [SRI](https://w3c.github.io/webappsec-subresource-integrity/).  Ideally, if SRI uses the same hash function, the value in the `Digest` header field is the same.

**Comment by LPardue** [comment]
+1, seems sensible to address this explicitly and help out implementers.

**Comment by ioggstream** [comment]
@martinthomson what do you think of the following?
We could define a new `digest-algorithm` named `sri`, which inherits the `sri` serialization,
eg.

```
<script src="https://example.com/example-framework.js"
        integrity="sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC"
        crossorigin="anonymous"></script>
```

could have it's related

```
GET /example-framework.js

HTTP/1.1 200 OK
Digest: sri=sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC

...whatever.js...

```

**Comment by martinthomson** [comment]
I don't think that would be a great outcome.  The idea is to have the oq... be generated the same way.

(For my benefit, because SRI and the whole "representation data" doesn't explicitly range requests or content codings, what is the state of play there?)

**Comment by ioggstream** [comment]
@martinthomson can you make an example of the expected request/response?

About encoding, probably in Sri case we should use an id-sha-384 to.cover the actual unencoded representation.

**Comment by LPardue** [comment]
From the SRI spec Section 3.4.5, Example 8 is a bit worrisome https://w3c.github.io/webappsec-subresource-integrity/#does-response-match-metadatalist

```
This algorithm allows the user agent to accept multiple, valid strong hash functions. 
For example, a developer might write a script element such as:

<script src="https://example.com/example-framework.js"
        integrity="sha384-Li9vy3DqF8tnTXuiaAJuML3ky+er10rcgNR/VqsVpcw+ThHmYcwiB1pbOxEbzJr7
                   sha384-+/M6kredJcxdsqkczBUjMLvqyHb1K/JThDXWsBVxMEeZHEaMKEOEct339VItX1zB"
        crossorigin="anonymous"></script>

which would allow the user agent to accept two different content payloads, 
one of which matches the first SHA384 hash value and the other matches the second SHA384 hash value.
````

SRI hashes are always calculated on the identity encoding, so AFAICT this example demonstrates an ability to hash two different representations and serve them. The Digest header doesn't quite work the same way.

**Comment by LPardue** [comment]
Regarding SRI and range requests, there is an open issue / feature request: https://github.com/w3c/webappsec-subresource-integrity/issues/57

**Comment by ioggstream** [comment]
I took some time to read https://w3c.github.io/webappsec-subresource-integrity/ 
and think about this.

## My position:

- our goal for this spec is to "fix" the inconsistencies related to representation manipulations and content-encoding;
- I'd avoid to entangle Digest to a complex spec like SRI.

This could be done:

- in a non-normative appendix;
- in another spec;


## SRI Complexities

### [algorithm agility](https://w3c.github.io/webappsec-subresource-integrity/#agility) 

>  the user agent will choose the strongest hash function in the list, 
>  via the https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction

This would be a breaking change in `Digest`, though we could suggest that `the receiver SHOULD use the strongest digest-algorithm`.


### [Multiple payloads](https://w3c.github.io/webappsec-subresource-integrity/#does-response-match-metadatalist)

The example posted by @LPardue will not be supported by Digest but:

- iiuc when the resource will be fetched, only one `Digest` value will be used

### SRI digest 

> digest https://w3c.github.io/webappsec-subresource-integrity/#digest

This seems incidentally compatible with the `digest-algorithm` encoding defined for `id-sha-256` and `id-sha-512`. The only missing digest-algorithm respect to SRI is `id-sha-384`.

Beware that this incidentality is not guaranteed to persist in future SRI implementation. At maximum we could state in an appendix or faq that `id-sha-*` is compatible with `sha*` from [REC-SRI-20160623](https://www.w3.org/TR/2016/REC-SRI-20160623/)

**Comment by LPardue** [comment]
@ioggstream said:

> This seems incidentally compatible with the digest-algorithm encoding defined for id-sha-256 and id-sha-512. The only missing digest-algorithm respect to SRI is id-sha-384.

> Beware that this incidentality is not guaranteed to persist in future SRI implementation. At maximum we could state in an appendix or faq that id-sha-* is compatible with sha* from REC-SRI-20160623

I don't think we can attempt to draw any lines between SRI and identity encoding due to [SRI Section 3.3](https://www.w3.org/TR/2016/REC-SRI-20160623/#apply-algorithm-to-response)

> 3.3 Response verification algorithms
> 3.3.1 Apply algorithm to response

>    Let result be the result of applying algorithm to the representation data without any content-codings applied, except when the user agent intends to consume the content with content-encodings applied. In the latter case, let result be the result of applying algorithm to the representation data.

**Comment by LPardue** [comment]
So up-levelling a bit here. 

Both Digest and SRI have what appears to be a similar goal but different implmentation characteristics. So perhaps it could help to capture this e.g. how does an implementer pick one, or how does it rationalise things if it chooses to use both. To @ioggstream point about where to do this, I think this it is probably best suited as a section in each document; however, they should acknowledge that neither requires the other.

SRI
> This specification defines a mechanism by which user agents may verify that a fetched resource has been delivered without unexpected manipulation.

Digest
> This document defines the Digest and Want-Digest header fields for HTTP, thus allowing client and server to negotiate an integrity checksum of the exchanged resource representation data.

By my reading SRI is more concerned with the integrity of resources delivered from a third party.  It allows the first-party to set client integrity expectations of a separately transferred resource; the expected digest is sent in one message (inside the message payload) and independently the resource is received in a different message. In effect, a client that reads SRI sets expectations on itself, with possibly some belief that the first party is more trustworthy than the third party. But of course, SRI can be used to load from first or third party.

In contrast, Digest is coupled to the message related to its representation. The sender informs the receiver of its integrity expectations i.e. "As best I can tell, this representation should have this hash". By combining Digest with Signature this can sort of be turned into "The signature party thinks the representation should have this hash".

Because SRI is carried in message payload, Digest can be used to protect the integrity of the integrity-check.

I think the correct way to square the circle of relationship between the two is to say that endpoints that implement these features should apply their requirements. And that when both SRI and Digest are supported, it makes sense to check Digest first. Why? Because digest can protect the transfer of a representation over multiple ranges, different paths and different content-encodings. Only after an endpoint has assurance that the composed representation has integrity within it's own self-defined terms should it then apply the integrity expectations set by a separate SRI party.

This means that Digest checks could pass but SRI could fail. I think that's fine.

**Comment by ioggstream** [comment]
@martinthomson reopening for clarity, as this was accidentally closed via the #971 PR

**Comment by LPardue** [comment]
I'm closing this due to inactivity and the fact that PR has landed that addresses the initial issue. If there are changes or additions to the current document, lets open new more specific issues.
